tools/project_manager/diff.php

```diff
@@ -176,46 +176,49 @@ function do_navigation($projectid, $image, $L_round_num, $R_round_num,
     $navigation_text .= "\n<input type='hidden' name='format' value='$format'>";
     $navigation_text .= "\n" . _("Jump to") . ": <select name='jumpto' onChange='$jump_to_js'>\n";
 
-    $condition = "";
-
-    if ($only_nonempty_diffs) {
-        $condition = "WHERE image = '$image' OR $L_text_column_name <> $R_text_column_name";
-    }
-
-    $query = "SELECT image, $L_user_column_name  FROM $projectid $condition ORDER BY image ASC";
+    $query = "SELECT image, $L_user_column_name, ($L_text_column_name = $R_text_column_name) AS is_empty_diff FROM $projectid ORDER BY image ASC";
```

Comment:
Not sure how expensive this is to run every time:

`($L_text_column_name = $R_text_column_name) AS is_empty_diff`

We only need the values calculated when `$only_nonempty_diffs` is set to true, so we could set it to something like `1 AS is_empty_diff` when it is false. That might be an unnecessary optimization and it would make the code less clear though.

Reply:
Why don't you run an explain on each version of the query, post the results here, and we can evaluate what the database will do to fulfill the query.

Reply:
Here are the results from the EXPLAIN command:

```
mysql> EXPLAIN SELECT image, round1_user, (master_text = round1_text) AS is_empty_diff FROM projectID463e6a9454485 ORDER BY image ASC;
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+----------------+
| id | select_type | table                  | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra          |
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+----------------+
|  1 | SIMPLE      | projectID463e6a9454485 | NULL       | ALL  | NULL          | NULL | NULL    | NULL |  608 |   100.00 | Using filesort |
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+----------------+
1 row in set, 1 warning (0.03 sec)
```

```
mysql> EXPLAIN SELECT image, round1_user, 0 AS is_empty_diff FROM projectID463e6a9454485 ORDER BY image ASC;
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+----------------+
| id | select_type | table                  | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra          |
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+----------------+
|  1 | SIMPLE      | projectID463e6a9454485 | NULL       | ALL  | NULL          | NULL | NULL    | NULL |  608 |   100.00 | Using filesort |
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+----------------+
1 row in set, 1 warning (0.00 sec)
```

I guess there is no meaningful difference? I ran it on this project: https://www.pgdp.org/~mlazaric/c/tools/project_manager/page_detail.php?project=projectID463e6a9454485&show_image_size=0 which has 608 pages and quite a few of the pages are 1000-2000 bytes.

Reply:
Yes, it's doing a tablescan for the query anyway, so they are largely equivalent.